domain/permission/state/state.go

```diff
@@ -195,25 +196,153 @@ func (s *State) UpsertPermission(ctx context.Context, args permission.UpsertPerm
 // ReadUserAccessForTarget returns the subject's (user) access for the
 // given user on the given target.
 func (s *State) ReadUserAccessForTarget(ctx context.Context, subject string, target corepermission.ID) (corepermission.UserAccess, error) {
-	return corepermission.UserAccess{}, errors.NotImplementedf("ReadUserAccessForTarget")
+	var userAccess corepermission.UserAccess
+	db, err := s.DB()
+	if err != nil {
+		return userAccess, errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		user, err := findUserByName(ctx, tx, subject)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		userAccess = user.toCoreUserAccess()
+
+		// Based on the grant to and grant from, find the permission,
+		// then the access type of it.
+		accessType, permissionUUID, err := findAccessType(ctx, tx, target.Key, user.UUID)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		userAccess.Access = corepermission.Access(accessType)
+		userAccess.PermissionID = permissionUUID
+		userAccess.Object = objectTag(target)
+		return nil
+	})
+	if err == nil {
+		return userAccess, nil
+	}
+	return corepermission.UserAccess{}, errors.Trace(domain.CoerceError(err))
```

Comment:
nit: flip this, do the error checking first.

Reply:
In many ways I dislike the pattern of:
```
if err != nil{
  return errors.Trace(err)
}
```
We're doing the `if err nil (or not nil)` check twice. Just return `errors.Trace(err)`. The first thing both errors.Trace and domain.CoerceError do is check err nil and return nil if it is.

Reply:
I'd rather
```
return userAccess, errors.Trace(domain.CoerceError(err))
```

Reply:
Personally, I'd rather be consistent with everywhere else. Return early on errors.